### PR TITLE
Enable undeploy-and-<do something> orders for deployed actors

### LINKED ISSUE #13039: Add support for undeploy on move order
All TS deployers need this to match original behavior. Might be wanted for RA/TD/D2k MCVs, too.

### PR DESCRIPTION
**Pull request by pchote**
This PR takes @tovl's hard work and all the things we learned during #16142 and iterates it several more steps towards hopefully being mergeable.

Fixes #13039.
I don't think this covers #16600 anymore - a separate PR will be needed to fix the transforms code.

Abbreviated history of the evolution between #16142 and now:

**Problem:** (from #16142) handing activities between actors is going to cause maintenance issues.
**Fix:** ~Introduce `IActivityTranferrableToTransformedActor` interface as suggested in the previous PR, which also required introducing a new `EnterTunnel` activity.~

**Problem:** Using alt to activate the undeploy orders clashes with `Passenger`'s alternate transport mode.
**Fix:** ~Change all cursors except Move to remain enabled while deployed.~ This had a nice side-effect of making `IWrapMove.CanWrapMoveOrder` mostly redundant, so I removed this and reimplemented the remaining move case using a condition.

**Problem:** Implementing `TransformIntoMobile` using a dummy `IMove` regressed targeting of construction yards by `AttackFollow` based units.
**Fix:** Reimplement deploy-to-transform behaviour without using `IMove`. This meant introducing a separate trait for each order type, but this downside was outweighed by supporting `EnterTunnel` and `Repairable` without having to rewrite their standard traits and activities. `IActivityTranferrableToTransformedActor` became redundant, and was replaced by a single concrete `IssueOrderAfterTransform` activity.

**Problem:** Further discussion about the cursor problem in Discord and #13039 made a reasonable argument that the alternate transport mode has been partly superseded by the newly fixed ability to queue multiple enter orders, and on balance it is better to sacrifice that feature than to bodge this one.
**Fix:** Removed AlternateTransportMode and restored the ability to require force-move to show the tunnel/repair/transport cursors.

If there is a consensus that this is the wrong call wrt AlternateTransportMode I can drop the last two commits - but please test the behaviour on the 4th vs 6th commits before arguing for this. IMO the current behaviour feels much nicer.

### CODE REVIEW

**Comment by pchote** [comment]
We're going to want a `TransformsIntoAircraft` to cover mod cases like the YR siege chopper.

**Review by PunkPun — COMMENTED** [review]
I did some testing in ra and ts, conyard undeploy is bugged. If tania places a c4 on a conyard and then it's undeployed using this method no animation will be played and conyard will instantly turn into an mcv. Same goes for chrono vortex.

**Comment by pchote** [comment]
The `TransformsInto*` traits need to check for `Transforms` being paused, not just disabled.

**Comment by tovl** [comment]
> I don't think this covers #16600 anymore

While I'm glad you were willing to take this over, this is a bit of a bummer, because it's arguably the more pressing issue.

I will review in more detail when I have time, but this is quite a busy week for me.

**Comment by pchote** [comment]
Indeed. However, with having to rewrite the transform-and-&lt;do something&gt; logic (twice) it was easier to drop back to the original `Transform` activity. Better to focus the fix in its own pr in any case, in case history repeats itself again here!

**Comment by pchote** [comment]
Fixed and updated.

**Review by PunkPun — COMMENTED** [review]
Checked out all mcv activities and it seems to work ok. There's just one thing worth discussing. Should the mcv still undeploy when the path to its destination is blocked?

**Review by abcdefg30 — DISMISSED** [review]
Looks good to me.

**Comment by pchote** [comment]
Fixed and changed `EnterAlliedActorTargeter` to use the same approach as #16509.

**Review by ghost — COMMENTED** [review]
Looks good, didn't spot any issues.

**Review by tovl — COMMENTED** [review]
- When queueing a move order after a deploy order on units with GrantConditionOnDeploy they will move in deployed state instead of undeploying:
Peek 2019-06-07 23-09. This is handled correctly in #16142 

- Orders (queued or otherwise) given to MCVs during the undeploy animation are ignored. This is not the case in #16142.

**Comment by pchote** [comment]
Fixed both issues.

File: OpenRA.Mods.Common/Traits/Carryall.cs [hunk 1]
```diff
@@ -355,23 +355,23 @@ public override bool CanTargetFrozenActor(Actor self, FrozenActor target, Target
...
 			readonly AircraftInfo aircraftInfo;
 			readonly CarryallInfo info;
-			readonly WVec carryableOffset;
+
+			public string OrderID { get { return "DeliverUnit"; } }
+			public int OrderPriority { get { return 6; } }
+			public bool IsQueued { get; protected set; }
+			public bool TargetOverridesSelection(TargetModifiers modifiers) { return true; }
 
 			public CarryallDeliverUnitTargeter(AircraftInfo aircraftInfo, CarryallInfo info, WVec carryableOffset)
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Carryall.cs]: `carryableOffset` appears to be unused now...?

**Review by ghost — COMMENTED** [review]
Can confirm the fixes for the queued orders for MCV and artillery.

File: OpenRA.Mods.Common/Traits/Carryall.cs (hunk 1 above)
Comment (pchote) [reply OpenRA.Mods.Common/Traits/Carryall.cs]: #16417 removed all uses of it. This just cleans up the unused variable as a part of the other changes.

Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/Carryall.cs]: But why is it still included in the targeter then?

Comment (pchote) [reply OpenRA.Mods.Common/Traits/Carryall.cs]: Ah, I see. Fixed.

**Review by reaperrr — APPROVED** [review]
Looks good to me and didn't encounter any issues

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/609c5a8738b1bc24e6f78b30118273dd234d29d6...2de24e1d70bef13b4ff997fdf963a6cbd0ec4dbe

**Comment by dragunoff** [comment]
There is a problem with "attack ground" - if a deployed unit is ordered to attack ground outside of firing range it would undeploy, move into range and just stop. 

I'm not sure what the behavior should be here. Either one of these:
- undeploy, move into range, deploy, attack ground
- undeploy, move into range, attack ground (if unit is able to attack in undeployd state)
- do not undeploy, ignore order if target is out of range

The last one seems most useful because the player probably would not want to undeploy when target is out of range (imagine a group of deployed artilery units). This however goes against how the force move modifier works.

**Comment by pchote** [comment]
An AttackDeploy to allow artillery to automatically deploy on attack orders has been discussed in other issues, and should come eventually in a later PR.

**Comment by dragunoff** [comment]
Yep, but as I said -- I wonder what the behavior should be. And also what about units that can attack in both states - like the tick tank.

**Comment by pchote** [comment]
This pr added a specific flag to disable attacking out of range targets on artillery and tick tanks, if this is breaking in some situations then that is a bug.

**Comment by tovl** [comment]
The problem with MCV orders being ignored during the undeploy animation wasn't quite fixed. Queued deploy orders are still ignored and move orders are always treated as queued even when shift is not pressed.

**Comment by matjaeck** [comment]
>  Queued deploy orders are still ignored

It works when you release the ALT key after the transformation is complete and then press the deploy key.

**Comment by GraionDilach** [comment]
> There is a problem with "attack ground" - if a deployed unit is ordered to attack ground outside of firing range it would undeploy, move into range and just stop.
> 
> I'm not sure what the behavior should be here. Either one of these:
> 
> * undeploy, move into range, deploy, attack ground
> * undeploy, move into range, attack ground (if unit is able to attack in undeployd state)
> * do not undeploy, ignore order if target is out of range
> 
> The last one seems most useful because the player probably would not want to undeploy when target is out of range (imagine a group of deployed artilery units). This however goes against how the force move modifier works.

This just emphasizes the issue that Attack Ground in it's current implementation is one giant UI hack and should have been transferred to a separate button within the command bar.

**Pull request merged**